### DISCUSSION TITLE
RuntimeWarning: Fitting failed on all retries.   warnings.warn("Fitting failed on all retries.", RuntimeWarning)

### DISCUSSION THREAD

**Discussion by LiuYang-1997** [meta-pytorch/botorch#1235] [Q&A] [answered]
Hello,
When I used botorch today, I found this warning, may I ask, what is the cause of this? Does this warning have any effect on the optimization process? And how to solve it?

**Comment by Balandat** [comment, marked as answer]
When we fit the surrogate models (i.e. optimize the MLL of the model), it is not unusual to run into numerical problems (typically due to ill-conditioned kernel matrices that can be caused by challenging training data in conjunction with the initial conditions of the hyperpriors on the parameters). If that's the case we usually re-sample starting points from the hyperpriors and retry the optimization. If that repeatedly fails this can indeed be a sign that the surrogate models are of poor quality, which will likely negatively impact the optimization performance. 

Can you provide some more detail on the problems setup, and ideally share a reproducible example? There are a few things to consider here - are you normalizing / standardizing your input data (the priors we have work best with data in the unit hypercube and outcomes that are zero mean unit variance)? Which model are you using? Are you inferring the noise level or providing noise observations?

**Reply by LiuYang-1997** [reply]
This warning occurred during my optimization of the Rosenbrock function using Expected Improvement and SingleTaskGP in a noiseless state.  I used lhs in pyDOE for sampling to generate the initial observations. Is the data not normalized the reason why this WARNING appears? Will data normalization improve the performance of the optimization? 
Thanks for your answer.

**Reply by Balandat** [reply]
> Is the data not normalized the reason why this WARNING appears?

It may well be. The prior on the lengthscales works best when the features are normalized to the unit (hyper)cube, and the prior on the noise is designed for observations that are standardized (zero mean, unit variance). If data is sort of in that ballpark that should be fine, but if not then this can lead to the kinds of numerical problems that you've observed. 

Botorch supports input and outcome transforms (https://github.com/pytorch/botorch/tree/main/botorch/models/transforms) that allow you to do this automatically in the background as part of the model. See e,g, the `initialize_model` function in https://github.com/pytorch/botorch/blob/main/tutorials/constrained_multi_objective_bo.ipynb for an example for how to use `Standardize`.

**Reply by LiuYang-1997** [reply]
Thank you. I will try to normalize the data to see if this warning will appear again.

**Reply by LiuYang-1997** [reply]
Thanks for your answer.
